tests/test_chain_db.py

```diff
@@ -0,0 +1,141 @@
+from regolith.chained_db import ChainDB
+
+
+def test_dddi():
+    a = {'a': {'a': {'a': 1}}}
+    z = ChainDB(a)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['a'], ChainDB)
+    assert isinstance(z['a']['a']['a'], int)
+    assert z['a']['a']['a'] + 1 == 2
+
+
+def test_second_mapping():
+    m1 = {'a': {'m': {'x': 0}}}
+    m2 = {'a': {'m': {'y': 1}}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'].maps, list)
+    assert z['a']['m']['y'] == 1
+
+
+def test_double_mapping():
+    m1 = {'a': {'m': {'y': 0}}}
+    m2 = {'a': {'m': {'y': 1}}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'].maps, list)
+    assert isinstance(z['a']['m']['y'], int)
+    assert z['a']['m']['y'] == 1
+
+
+def test_list_mapping():
+    m1 = {'a': {'m': 'x'}}
+    m2 = {'a': {'m': 'y'}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], str)
+    assert z['a']['m'] == 'y'
+
+
+def test_mixed_mapping():
+    m1 = {'a': {'m': {'y': 1}}}
+    m2 = {'a': {'m': 1}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], int)
+    assert z['a']['m'] == 1
+
+
+def test_exactness():
+    d = {'y': 1}
+    m1 = {'a': {'m': d}}
+    z = ChainDB(m1)
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], ChainDB)
+    assert isinstance(z['a']['m'].maps[0], dict)
+    assert d is z['a']['m'].maps[0]
+
+
+def test_exactness_setting():
+    d = {'y': 1}
+    m1 = {'a': {'m': d}}
+    z = ChainDB(m1)
+    e = {'z': 2}
+    z['a']['m'] = e
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], ChainDB)
+    assert isinstance(z['a']['m'].maps[0], dict)
+    assert e is z['a']['m'].maps[0]
+
+
+def test_exactness_setting_multi():
+    d = 'a'
+    e = 'b'
+    m1 = {'a': {'m': d}}
+    m2 = {'a': {'m': e}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    g = 'c'
+    z['a']['m'] = g
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], str)
+    assert isinstance(z['a'].maps[0], dict)
+    assert g is z['a'].maps[1]['m']
+    # We sent this to the first map not the last
+    assert z['a']['m'] is g
+
+
+def test_exactness_setting_multi2():
+    d = [1, 2]
+    e = [3, 4]
+    ee = [5, 6]
+    m1 = {'a': {'m': d}}
+    m2 = {'a': {'m': e, 'mm': ee}}
+    z = ChainDB(m1)
+    z.maps.append(m2)
+    g = [-1, -2]
+    z['a']['mm'] = g
+    assert isinstance(z['a'], ChainDB)
+    assert isinstance(z['a']['m'], list)
+    assert isinstance(z['a'].maps[0], dict)
+    assert g is z['a'].maps[1]['mm']
+    # We sent this to the first map not the last
+    assert z['a']['m'] is not g
```

Comment:
Was this supposed to be `z['a']['mm']`? I am not sure that I understand the comment. 

Also did g go to m2 because m2 already had an `'mm'` key or because m2 is the last entry in the maps list?

Reply:
from the following test it looks like it is because it m2 had 'mm'

Reply:
Correct, we first look through all the elements in the `ChainedDB` to see if any of the keys match. If there is a match we overwrite else we send it to the first map.